doc/_extensions/vuln_table.py

```diff
@@ -0,0 +1,133 @@
+from docutils import nodes
+from sphinx.util.docutils import SphinxDirective
+from sphinx.application import Sphinx
+from azure.storage.blob import ContainerClient
+from azure.core.pipeline import PipelineResponse
+from sphinx.util import logging
+from sphinx.util.console import bold, term_width_line
+from pathlib import Path
+import requests
+import json
+import pickle
+
+AZ_CONN_STR_PUBLIC = ";".join(
+    (
+        "DefaultEndpointsProtocol=https",
+        "EndpointSuffix=core.windows.net",
+        "AccountName=ncsdocsa",
+    )
+)
+"""Azure connection string (public acces)."""
+
+AZ_CONTAINER = "ncs-doc-vuln"
+"""Azure container."""
+
+logger = logging.getLogger(__name__)
+
+class VulnTable(SphinxDirective):
+
+    def run(self):
+        if not hasattr(self.env, "vuln_cache"):
+            return [nodes.paragraph(text="No vulnerabilities found")]
+
+        cache = self.env.vuln_cache
+        keys = cache["schema"]
+        table = nodes.table()
+        tgroup = nodes.tgroup(cols=len(keys))
+        for _ in range(len(keys)):
+            colspec = nodes.colspec(colwidth=1)
+            tgroup.append(colspec)
+        table += tgroup
+
+        # table head row
+        thead = nodes.thead()
+        tgroup += thead
+        row = nodes.row()
+        for key in keys:
+            entry = nodes.entry()
+            entry += nodes.paragraph(text=key)
+            row += entry
+        thead.append(row)
+
+        # table body
+        rows = []
+        for vuln in cache["body"]:
+            row = nodes.row()
+            rows.append(row)
+
+            for key in keys:
+                entry = nodes.entry()
+                para = nodes.paragraph(text=vuln[key])
+                entry += para
+                row += entry
+
+        tbody = nodes.tbody()
+        tbody.extend(rows)
+        tgroup += tbody
+
+        return [table]
+
+def progress(response: PipelineResponse) -> None:
+    """Download progress callback."""
+
+    mb_tot = response.context["data_stream_total"] / (1024 * 1024)
+    mb_curr = response.context["download_stream_current"] / (1024 * 1024)
+    pct = int((mb_curr / mb_tot) * 100)
+    s = term_width_line(
+        bold("Downloading vulnerability files... ")
+        + f"{mb_curr:.2f} of {mb_tot:.2f} Mb [{pct} %]"
+    )
+    logger.info(s, nonl=pct < 100)
+
+def download_vuln_table(app: Sphinx) -> None:
+    """Download vulnerability table from cache"""
+
+    # Check if local cache exists
+    try:
+        with open(Path(app.doctreedir) / Path("vuln_cache.pickle"), "rb") as cache_file:
+            local_cache = pickle.load(cache_file)
+    except FileNotFoundError:
+        logger.info("No vulnerability cache found locally")
+        local_cache = None
```

Comment:
note: instead of creating a pickle file, you can use the Sphinx environment (which is preserved across builds). See for example https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/_extensions/zephyr/doxyrunner.py#L232 Cache extension is a bit more complicated (because we need to patch the environment), I think you don't need to do anything like that here